### PR TITLE
Move include/datetime to src - Part 1

### PR DESCRIPTION
**Pull request by MrPetovan**
Part of #3878

I started moving functions in `include/datetime.php`, but I had to stop at `datetime_convert()` because of how frequent it is.

Moreover, I finally got the chance of changing the parameter order for something that makes more sense regarding the default values.

The part 2 will deal with the rest of the functions and will probably be lighter.

This is currently running on my node with success.

This PR is also a way of showcasing my method for moving functions from `include` to `src`:
1. I create the class in `src` with all the methods copied from the `include` file, renamed if needed
2. I empty the functions in the `include` file to only return the result from the newly created class methods.
3. I look for all instances of each function in the project folder and I replace by the class method name. NetBeans provides a tool to do that, but I still do a string search for callback functions. I check `use` statements.
4. Once I replaced all the instances of the old function, I can delete it from the include file and move on to the next. This way I always have a working project even in the middle of my refactoring.
5. Once I deleted all the functions from the `include` file, I remove all the references to the files.
6. Last step, the most gratifying, deleting the `include` file. Move complete.

### CODE REVIEW

File: mod/contacts.php [hunk 1]
```diff
@@ -554,7 +554,7 @@ function contacts_content(App $a)
 		// tabs
 		$tab_str = contacts_tab($a, $contact_id, 2);
 
-		$lost_contact = (($contact['archive'] && $contact['term-date'] > NULL_DATE && $contact['term-date'] < datetime_converL10n::t('', '', 'now')) ? L10n::t('Communications lost with this contact!') : '');
```
Comment (annando) [inline mod/contacts.php]: Ups ... have you had a look at other possible false replacements of ```t(```?

Comment (MrPetovan) [reply mod/contacts.php]: How do you suggest I check this when it isn't failing?

Comment (zeroadam) [reply mod/contacts.php]: This **should** be the only one that slipped through the cracks.. I was as careful as I could be when searching through the million occurrences of t(

**Comment by annando** [comment]
I currently do have problems reviewing this PR. Not only that github doesn't load all differences, but also I currently have some issues with my hand. So it would be great if someone else could review it.

**Comment by MrPetovan** [comment]
I'm sorry to hear that.

From any reviewer: I suggest you go a commit at a time if the whole diff is too long. I made an effort to commit atomic changes at each step of the way, so you should be able to see what went on organically.

**Comment by annando** [comment]
Watching your commits, it seems as if this is "work in progress"?

**Comment by MrPetovan** [comment]
I'm just fiddling with explicit shorthands for `Temporal::convert()`. Overall, this is a WIP as it is part 1 of the move. If you want the whole enchilada in one go I can do that as well 😄

**Comment by zeroadam** [comment]
I have been trying to put some time in on this, especially to relive @annando. However, that's like having a Junior peer review a Senior's code :rofl:

**Comment by MrPetovan** [comment]
I appreciate the compliment, and if you have any questions about this PR, code or philosophy, please don’t hesitate to ask.

**Comment by zeroadam** [comment]
Certainly! At work I'm a "mid-level" and we often make the reference you can't have a junior reviewing a mid-level, in regard to our junior devs. But here, with php and this project, I'm definitely the new kid on the block! :smile:

**Comment by MrPetovan** [comment]
On the contrary, reviewing the work of someone more experienced is both a great way to learn and a chance to spot something the seansoned dev will more easily look over.

**Comment by zeroadam** [comment]
Ooh yes, agree. In our case it is meant as the single approving review.. our juniors participate in the reviews but not as a single reviewer.

**Comment by MrPetovan** [comment]
Oh, I better understand. It’s the same deal here though, you can’t merge yourself yet.

**Comment by MrPetovan** [comment]
Ok, I think I'm done for Part 1 😅

Part 2 will be way more relaxing, I promise.

**Comment by MrPetovan** [comment]
This has been painfully rebased against the latest function moves.

File: boot.php [hunk 2]
```diff
@@ -21,16 +21,15 @@
...
+use Friendica\Core\L10n;
 use Friendica\Core\PConfig;
+use Friendica\Core\System;
 use Friendica\Core\Worker;
 use Friendica\Database\DBM;
-use Friendica\Model\Contact;
 use Friendica\Database\DBStructure;
-use Friendica\Module\Login;
+use Friendica\Model\Contact;
+use Friendica\Util\DateTimeFormat;
```
Comment (tobiasd) [inline boot.php]: Just curious, is there a reasoning behind this constantly re-sorting of the `use` statements?

File: src/Util/Temporal.php [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+				$input_text,
+				'',
+				$required ? '*' : '',
+				'placeholder="' . $readable_format . '"'
+			],
+		]);
+
+		$o .= "<script type='text/javascript'>";
+		$o .= "\$(function () {var picker = \$('#id_$id').datetimepicker({step:5,format:'$dateformat' $minjs $maxjs $pickers $defaultdatejs}); $extra_js})";
+		$o .= "</script>";
```
Comment (tobiasd) [inline src/Util/Temporal.php]: This could probably be moved to the template as well.

File: src/Util/Temporal.php [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+	 *
+	 * @param int $y Year
+	 * @param int $m Month (1=January, 12=December)
+	 *
+	 * @return int Number of days in the given month
+	 */
+	public static function getDaysInMonth($y, $m)
+	{
+		return date('t', mktime(0, 0, 0, $m, 1, $y));
+		;
```
Comment (tobiasd) [inline src/Util/Temporal.php]: stray semicolon?

File: boot.php (hunk 2 above)
Comment (MrPetovan) [reply boot.php]: I have an automatic tool that checks for class usage inside a file and suggest a use list alphabetically sorted. Not everyone use this tool, and sometimes people add uses that aren't alphabetically sorted. If I'm subsequently editing the file and using my IDE tool, it sorts the use statements back.

File: src/Util/Temporal.php (hunk 3 above)
Comment (MrPetovan) [reply src/Util/Temporal.php]: Not really, the `field_input.tpl` template is really generic, so a specific bit of Javascript about a datetimepicker wouldn't be welcome. Does it deserve its own template? Probably, but I'm not in a rush to add it.

File: src/Util/Temporal.php (hunk 4 above)
Comment (MrPetovan) [reply src/Util/Temporal.php]: Indeed, nice catch.

**Comment by MrPetovan** [comment]
This has been painfully rebased onto the latest function moves.

**Comment by annando** [comment]
I'm currently at the fosdem. I hope that @tobiasd or @rabuzarus will review and accept it. (Especially since @zeroadam already did).

**Comment by tobiasd** [comment]
> I'm currently at the fosdem. I hope that @tobiasd or @rabuzarus will
> review and accept it. (Especially since @zeroadam already did).

I went through it and remarked things that I think should be changed. I'll have a look later to see if they are and then merge.

File: src/Util/Temporal.php (hunk 3 above)
Comment (MrPetovan) [reply src/Util/Temporal.php]: Ok, I have to create a specific template for this field then.

Comment (MrPetovan) [reply src/Util/Temporal.php]: Done.

**Comment by MrPetovan** [comment]
Don't forget https://github.com/friendica/friendica-addons/pull/501

**Pull request merged**